File: elixir/lib/merge_two_sorted_lists.ex
defmodule MergeTwoSortedLists do
  @moduledoc """
  See: <https://leetcode.com/problems/merge-two-sorted-lists/>
  """

  @spec merge_two_lists(
          list1 :: ListNode.t() | nil,
          list2 :: ListNode.t() | nil
        ) :: ListNode.t() | nil
  def merge_two_lists(left, nil), do: left
  def merge_two_lists(nil, right), do: right

  def merge_two_lists(left, right) do
    if left.val < right.val do
      %ListNode{val: left.val, next: merge_two_lists(left.next, right)}
    else
      %ListNode{val: right.val, next: merge_two_lists(left, right.next)}
    end
  end
end

File: javascript/src/merge_two_sorted_lists.js
/**
 * @see <https://leetcode.com/problems/merge-two-sorted-lists/>
 *
 * @param {(ListNode | null)} left
 * @param {(ListNode | null)} right
 * @return {ListNode}
 */
function mergeTwoLists(left, right) {
    if (left === null) {
        return right;
    } else if (right === null) {
        return left;
    } else if (left.val < right.val) {
        return new ListNode(left.val, mergeTwoLists(left.next, right));
    } else {
        return new ListNode(right.val, mergeTwoLists(left, right.next));
    }
}

class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

export {mergeTwoLists, ListNode};
